### PR TITLE
Add tests for RollupFunnel + make it stable

### PR DESCRIPTION
**Pull request by trentmwillis**
Addressing #256 & #257 

cc @stefanpenner @nathanhammond @rwjblue

### CODE REVIEW

File: node-tests/rollup-funnel-test.js [hunk 1]
```diff
@@ -0,0 +1,93 @@
...
+          external: [ 'ember-engines/routes' ]
+        }
+      });
+
+      var builder = new broccoli.Builder(rollupFunnel);
+
+      return builder.build().then(function(result) {
+        var output = walkSync(result.directory);
+        assert.deepEqual(output, [ 'engine.js', 'utils/', 'utils/herp.js' ]);
+        return builder.cleanup();
```
Comment (stefanpenner) [inline node-tests/rollup-funnel-test.js]: should be called in `afterEach`

likely in the category of "an exploration of an abstraction to hide these details"

Comment (stefanpenner) [inline node-tests/rollup-funnel-test.js]: https://github.com/chadhietala/broccoli-rollup/blob/master/tests/index.js#L18
https://github.com/chadhietala/broccoli-rollup/blob/master/tests/index.js#L40
https://github.com/chadhietala/broccoli-rollup/blob/master/tests/index.js#L35

Comment (trentmwillis) [inline node-tests/rollup-funnel-test.js]: Generally, I dislike that pattern as it encourages state sharing, which can lead to test ordering problems. In this case, however, it is likely okay. Will just also suggest setting `builder` to `undefined` in `afterEach`.

**Comment by trentmwillis** [comment]
@stefanpenner updated with your comments.

**Comment by trentmwillis** [comment]
Updated this with an attempt at making rebuilds stable/performant. Not quite finished though.

These are the scenarios I am thinking of:

- No changes -> output is stable
- Using `include` and dep graph changes -> run rollup to recalculate output
- Using `include` and change occurs outside dep graph -> output is stable
- Using `exclude` and dep graph changes -> run rollup to recalculate output
- Using `exclude` and change occurs outside dep graph -> copy forward changes

File: lib/rollup-funnel.js [hunk 2]
```diff
@@ -45,25 +47,95 @@ function RollupFunnel(inputNodes, options) {
...
+    var absolutePath = path.join(inputPath, entryPath);
+
+    if (!fs.existsSync(absolutePath)) {
+      return;
+    }
+
+    var stat = fs.lstatSync(absolutePath);
+    var isDirectory = stat.isDirectory();
+
+    return {
```
Comment (trentmwillis) [inline lib/rollup-funnel.js]: I would appreciate this, or even having `fromPaths` do the stating. It was super unintuitive to me why `fromPaths` wasn't doing what I wanted it to do.

File: lib/rollup-funnel.js [hunk 3]
```diff
@@ -45,25 +47,95 @@ function RollupFunnel(inputNodes, options) {
...
+        }, this);
+        return;
+      }
+    }
+  }
+
   var modules = [];
 
-  var entryExists = existsSync(path.join(this.inputPaths[0], this.options.rollup.entry));
+  var entryExists = existsSync(path.join(base, this.options.rollup.entry));
```
Comment (trentmwillis) [inline lib/rollup-funnel.js]: Not intentional, just due to me overlooking it. I think I added all the `fs.existsSync`, so I'll update to use `exists-sync` instead.

File: lib/rollup-funnel.js [hunk 4]
```diff
@@ -99,16 +171,26 @@ RollupFunnel.prototype.build = function() {
   return rollup(rollupOptions).then(function() {
     var toCopy;
 
+    this._depGraph = modules.sort();
+    this._nonDepGraph = filter(base, '', function(module) {
+      return !~modules.indexOf(module);
```
Comment (trentmwillis) [inline lib/rollup-funnel.js]: Nah, I'll change it.

**Comment by trentmwillis** [comment]
@rwjblue addressed most of your comments. I think this is in a pretty good place now.

**Comment by rwjblue** [comment]
> Also, would we potentially want to extract this into a separate package? Would be nice to iterate on it independent of the ember-engines release cycle

Yes, I agree.

**Comment by trentmwillis** [comment]
@rwjblue that was my plan

**Comment by rwjblue** [comment]
This looks good to me, I'd like to land this and release so that we can continue to iterate (and possibly extract). 

Any objections?

**Pull request merged**